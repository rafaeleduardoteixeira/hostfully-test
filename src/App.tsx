import React from 'react';
import { createBrowserRouter, RouterProvider } from 'react-router-dom';
import createdRoutes from './constants/router';
import { Header } from './components/Header/Header';
import { AppContainer } from './App.styles';

function App() {
  return (
    <AppContainer>
      <Header />
      <RouterProvider router={createdRoutes} />
    </AppContainer>
  );
}

export default App;
