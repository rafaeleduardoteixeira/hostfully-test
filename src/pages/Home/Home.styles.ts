import styled from 'styled-components';
import { colors, media } from '../../themes';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
`;

export const HomeBanner = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  position: relative;
`;

export const TitleBanner = styled.h1`
  position: absolute;
  color: white;
  font-size: 2.5rem;
  font-weight: 700;

  @media ${media.mobile} {
    font-size: 1%.8;
    text-align: center;
  }
`;

export const ImgBanner = styled.div`
  background-image: url('/images/banner.jpeg');
  border-radius: 10px;
  background-size: cover;
  height: 450px;
  width: 100%;

  @media ${media.mobile} {
    border-radius: 10px;
    height: 300px;
    background-image: url('/images/banner-mobile.jpeg');
  }
`;

export const SearchContainer = styled.div`
  position: absolute;
  bottom: -40px;
  width: 50%;
  height: 80px;
  background-color: white;
  border-radius: 10px;
  box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, 0.1);

  display: flex;
  align-items: center;
  justify-content: space-between;

  padding: 0 20px;

  @media ${media.tablet} {
    width: 70%;
  }

  @media ${media.mobile} {
    height: 60px;
    width: 90%;
    padding: 0 10px;
    bottom: -30px;
  }
`;

export const SearchButton = styled.button`
  height: 45px;
  width: 45px;
  background-color: ${colors.hostfullyPrimaryColor};
  border: none;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 50%;

  margin-left: 20px;

  @media ${media.mobile} {
    width: 50px;
    margin-left: 10px;
  }

  &:hover {
    background-color: ${colors.hostfullySecondaryColor};
    transition: 0.3s;
  }
`;

export const SearchIcon = styled.img`
  width: 22px;
  height: 22px;
`;
