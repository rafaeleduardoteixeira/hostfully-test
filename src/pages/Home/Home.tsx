import { useContext, useEffect, useState } from 'react';
import {
  Container,
  HomeBanner,
  ImgBanner,
  TitleBanner,
  SearchContainer,
  SearchButton,
  SearchIcon,
} from './Home.styles';
import { PropertyList } from '../../components/PropertyList/PropertyList';
import { SearchInput } from '../../components/SearchInput/SearchInput';
import { AppContext, AppContextType } from '../../Context/App.context';
import { Property } from '../../interfaces/property';

export const Home = (): JSX.Element => {
  const { isLoading, getProperties } = useContext(AppContext) as AppContextType;
  const [properties, setProperties] = useState<Property[]>([]);
  const [searchTerm, setSearchTerm] = useState('');

  useEffect(() => {
    setProperties(getProperties());
  }, [getProperties]);

  const handleChangeSearchTerm = (value: string) => {
    setSearchTerm(value);
  };

  // This function is responsible to filter the properties based on the search term
  const handleSearch = () => {
    setProperties(getProperties(searchTerm));
  };

  return (
    <Container>
      <HomeBanner>
        <TitleBanner>Find Your Property to Rental</TitleBanner>
        <ImgBanner />
        <SearchContainer>
          <SearchInput onChange={handleChangeSearchTerm} />
          <SearchButton onClick={() => handleSearch()}>
            <SearchIcon src="/icons/search.svg" alt="search" title="Search an property" />
          </SearchButton>
        </SearchContainer>
      </HomeBanner>
      <PropertyList isLoading={isLoading} properties={properties} />
    </Container>
  );
};
