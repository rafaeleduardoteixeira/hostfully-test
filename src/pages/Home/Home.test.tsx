import React from 'react';
import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { Home } from './Home';
import { AppContext } from '../../Context/App.context';

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
}));

describe('Home', () => {
  test('Snapshot test', () => {
    const { asFragment } = render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <Home />
      </AppContext.Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('Should render Home', () => {
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <Home />
      </AppContext.Provider>,
    );
    const HomeElement = screen.getByTestId('Home');
    expect(HomeElement).toBeInTheDocument();
  });
  test('Should type the search input', () => {
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <Home />
      </AppContext.Provider>,
    );
    const InputSearchElement = screen.getByPlaceholderText('Enter a property name or city');
    fireEvent.change(InputSearchElement, { target: { value: 'search' } });
    expect(InputSearchElement).toHaveValue('search');
  });
  test('Should click the search button', () => {
    const getProperties = jest.fn();
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties,
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <Home />
      </AppContext.Provider>,
    );
    const InputSearchElement = screen.getByPlaceholderText('Enter a property name or city');
    fireEvent.change(InputSearchElement, { target: { value: 'search' } });
    const SearchButtonElement = screen.getByTestId('SearchButton');
    fireEvent.click(SearchButtonElement);
    expect(getProperties).toBeCalled();
  });
});
