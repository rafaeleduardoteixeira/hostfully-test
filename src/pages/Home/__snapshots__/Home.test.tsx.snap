// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="sc-braxZu kWsybE"
    data-testid="Home"
  >
    <div
      class="sc-gJhJTp hvkuey"
    >
      <h1
        class="sc-elDIKY jBsRBE"
      >
        Find Your Property to Rental
      </h1>
      <div
        class="sc-fQpRED gIEbba"
      />
      <div
        class="sc-dsLQwm icBkyY"
      >
        <div
          class="sc-hfvVTD gMNIZl"
          data-testid="SearchInput"
        >
          <input
            class="sc-ifyrTC hnIWaB"
            placeholder="Enter a property name or city"
          />
        </div>
        <button
          class="sc-iKTcqh bWQgxY"
          data-testid="SearchButton"
        >
          <img
            alt="search"
            class="sc-gnpbhQ iceGBX"
            src="/icons/search.svg"
            title="Search an property"
          />
        </button>
      </div>
    </div>
    <div
      class="sc-la-DxNn gCDkMC"
      data-testid="PropertyList"
    >
      <h1
        class="sc-iCZwEW ifeOmH"
      >
        Properties For Rent
      </h1>
      <div
        class="sc-knesRu fjeQcU"
      />
      <div
        class="sc-dnaUSb cXawOW"
      />
    </div>
  </div>
</DocumentFragment>
`;
